scripts/lib/CIME/case/case_setup.py

```diff
@@ -152,10 +152,15 @@ def _case_setup_impl(case, caseroot, clean=False, test_mode=False, reset=False):
 
             case.initialize_derived_attributes()
 
-            cost_per_node = 16 if case.get_value("MACH") == "yellowstone" else case.get_value("MAX_MPITASKS_PER_NODE")
+            cost_per_node = case.get_value("MAX_MPITASKS_PER_NODE")
             case.set_value("COST_PES", case.num_nodes * cost_per_node)
-            case.set_value("TOTALPES", case.total_tasks)
-            case.set_value("SMP_PRESENT", case.get_build_threaded())
+            threaded = case.get_build_threaded()
+            case.set_value("SMP_PRESENT", threaded)
+            if threaded and case.total_tasks * case.thread_count > cost_per_node:
+                smt_factor = max(1.0,int(case.get_value("MAX_TASKS_PER_NODE") / cost_per_node))
+                case.set_value("TOTALPES", int(case.total_tasks * max(1.0,float(case.thread_count) / smt_factor)))
+            else:
+                case.set_value("TOTALPES", case.total_tasks*case.thread_count)
```

Comment:
Hyperthreading degree is based on MAX_TASKS_PER_NODE / MAX_MPITASKS_PER_NODE. On e3sm KNLs, hyper-threading factor is usually 1, 2 or 4, depending on the chosen pelayout. In the example above, the 2 OMP threads are assigned to 2 out of 4 available hyper-threads on a core for a total of 16 actual cores.

Reply:
If you are using less that a full node - less that 64 total  tasks on knl then you should not hyperthread.  If you are you are wasting resources.

Reply:
I think that whether we are or are not hyperthreading on knl when using less than a full node is going a bit into the weeds (doesn't matter).    If this pr is otherwise okay lets let this particular issue go.